src/AstGen.zig

```diff
@@ -10130,14 +10128,21 @@ fn detectLocalShadowing(
     };
 }
 
-fn advanceSourceCursor(astgen: *AstGen, source: []const u8, end: usize) void {
+fn advanceSourceCursor(astgen: *AstGen, end: ast.ByteOffset) void {
     var i = astgen.source_offset;
     var line = astgen.source_line;
     var column = astgen.source_column;
+    if (end < i) {
+        // there are some cases where this function is not monotonically through the source
+        // in those cases, reset i back to the beginning to ensure we get a correct result
+        i = 0;
+        line = 1;
+        column = 1;
+    }
```

Comment:
Instead of detecting data being in an invalid state and falling back to the the O(N^2) algorithm, we need to assert that data will never be in an invalid state, and then go figure out how that was happening. It may require reworking a bit of logic.